third_party/terraform/resources/resource_logging_bucket_config.go

```diff
@@ -0,0 +1,183 @@
+	},
+	"bucket_id": {
+		Type:     schema.TypeString,
+		Required: true,
+		ForceNew: true,
+	},
+	"description": {
+		Type:     schema.TypeString,
+		Optional: true,
+		Computed: true,
+	},
+	"retention_days": {
+		Type:     schema.TypeInt,
+		Optional: true,
+		Default:  30,
+	},
+	"lifecycle_state": {
+		Type:     schema.TypeString,
+		Computed: true,
+	},
+}
+
+type loggingBucketConfigIDFunc func(d *schema.ResourceData, config *Config) (string, error)
+
+// ResourceLoggingBucketConfig creates a resource definition by merging a unique field (eg: folder) to a generic logging bucket
+// config resource. In practice the only difference between these resources is the url location.
+func ResourceLoggingBucketConfig(parentType string, parentSpecificSchema map[string]*schema.Schema, iDFunc loggingBucketConfigIDFunc) *schema.Resource {
+	// func ResourceLoggingBucketConfig() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceLoggingBucketConfigAcquire(iDFunc),
+		Read:   resourceLoggingBucketConfigRead(),
+		Update: resourceLoggingBucketConfigUpdate(),
+		Delete: resourceLoggingBucketConfigDelete(),
+		Importer: &schema.ResourceImporter{
+			State: resourceLoggingBucketConfigImportState(parentType),
+		},
+		Schema: mergeSchemas(loggingBucketConfigSchema, parentSpecificSchema),
+	}
+}
+
+var loggingBucketConfigIDRegex = regexp.MustCompile("(.+)/(.+)/locations/(.+)/buckets/(.+)")
+
+func resourceLoggingBucketConfigImportState(parent string) schema.StateFunc {
+	return func(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+		// loggingSinkId, err := parseLoggingSinkId()
+		parts := loggingBucketConfigIDRegex.FindStringSubmatch(d.Id())
+		if parts == nil {
+			return nil, fmt.Errorf("unable to parse logging sink id %#v", d.Id())
+		}
+
+		if len(parts) != 5 {
+			return nil, fmt.Errorf("Invalid id format. Format should be '{{parent_type}}/{{parent_id}}/locations/{{location}}/buckets/{{bucket_id}} with parent_type in %s", loggingSinkResourceTypes)
+		}
+
+		validLoggingType := false
+		for _, v := range loggingSinkResourceTypes {
+			if v == parts[1] {
+				validLoggingType = true
+				break
+			}
+		}
+		if !validLoggingType {
+			return nil, fmt.Errorf("Logging parent type %s is not valid. Valid resource types: %#v", parts[1],
+				loggingSinkResourceTypes)
+		}
+
+		d.Set(parent, parts[1]+"/"+parts[2])
+
+		d.Set("location", parts[3])
+
+		d.Set("bucket_id", parts[4])
+
+		return []*schema.ResourceData{d}, nil
+	}
+}
+
+func resourceLoggingBucketConfigAcquire(iDFunc loggingBucketConfigIDFunc) func(*schema.ResourceData, interface{}) error {
+	return func(d *schema.ResourceData, meta interface{}) error {
+		config := meta.(*Config)
+
+		id, err := iDFunc(d, config)
+		if err != nil {
+			return err
+		}
+
+		d.SetId(id)
+
+		return resourceLoggingBucketConfigUpdate()(d, meta)
+	}
+}
+
+func resourceLoggingBucketConfigRead() schema.ReadFunc {
+	return func(d *schema.ResourceData, meta interface{}) error {
```

Comment:
Is this just returning a ReadFunc (instead of being a ReadFunc) for consistency with create, or is there a technical reason why it does that? Can you add a comment explaining?

Reply:
It was mostly for consistency, but even that now seems like a weak argument. I'll just change them to direct.